src/sql/workbench/services/notebook/common/notebookRegistry.ts

```diff
@@ -94,16 +95,67 @@ let notebookContrib: IJSONSchema = {
 		}
 	]
 };
+let notebookLanguageMagicType: IJSONSchema = {
+	type: 'object',
+	default: { magic: '', language: '', kernels: [], executionTarget: null },
+	properties: {
+		magic: {
+			description: localize('carbon.extension.contributes.notebook.magic', 'Name of the cell magic, such as "%%sql".'),
+			type: 'string'
+		},
+		language: {
+			description: localize('carbon.extension.contributes.notebook.language', 'What language should the cell be if this cell magic is included?'),
+			type: 'string'
+		},
+		executionTarget: {
+			description: localize('carbon.extension.contributes.notebook.executionTarget', 'Optional execution target this magic indicates, for example Spark vs SQL'),
+			type: 'string'
+		},
+		kernels: {
+			description: localize('carbon.extension.contributes.notebook.kernels', 'Optional set of kernels this is valid for, e.g. python3, pyspark3, sql'),
+			oneOf: [
+				{ type: 'string' },
+				{
+					type: 'array',
+					items: {
+						type: 'string'
+					}
+				}
+			]
+		}
+	}
+};
+
+let languageMagicContrib: IJSONSchema = {
+	description: localize('vscode.extension.contributes.notebook.languagemagics', "Contributes notebook language."),
+	oneOf: [
+		notebookLanguageMagicType,
+		{
+			type: 'array',
+			items: notebookLanguageMagicType
+		}
+	]
+};
+
+export interface NotebookLanguageMagicRegistration {
+	magic: string;
+	language: string;
+	kernels?: string[];
```

Comment:
are there some magics that don't apply to kernels?

Reply:
There could be, yes